tikv cluster restart failed with opaque error

Please answer these questions before submitting your issue. Thanks!

1. What version of Rust are you using (`rustc --version`)?
used pingcap/tikv image

2. What operating system and processor architecture are you using?
centos7/amd64

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
tikv cluster restart

4. What did you expect to see?
with consistent storage etc, tikv cluster shall be able to restart 

5. What did you see instead?
restart failed for all 3 tikv instance with following error:

```
2017/03/09 13:30:54.087 mod.rs:375: [INFO] Welcome to TiKV.
2017/03/09 13:30:54.087 mod.rs:376: [INFO] Version:
2017/03/09 13:30:54.087 mod.rs:377: [INFO] Git Commit Hash: a64a0af172b53b9896c3ff7cc3ed5b24a4938b16
2017/03/09 13:30:54.087 mod.rs:378: [INFO] UTC Build Time:  2017-02-27 01:45:21
2017/03/09 13:30:54.087 mod.rs:379: [INFO] Rustc Version:   1.17.0-nightly (ba7cf7cc5 2017-02-11)
2017/03/09 13:30:54.087 tikv-server.rs:160: [INFO] rocksdb.max-open-files use default Some(40960)
2017/03/09 13:30:54.088 tikv-server.rs:251: [WARN] Limit("kernel parameters net.core.somaxconn got 128, expect 32768")
2017/03/09 13:30:54.088 tikv-server.rs:251: [WARN] Limit("check_kernel_params failed No such file or directory (os error 2)")
2017/03/09 13:30:54.088 tikv-server.rs:251: [WARN] Limit("kernel parameters vm.swappiness got 60, expect 0")
2017/03/09 13:30:54.088 tikv-server.rs:97: [INFO] flag A: Some("0.0.0.0:20160")
2017/03/09 13:30:54.088 tikv-server.rs:97: [INFO] flag pd: Some("pd-0.tipd:2379,pd-1.tipd:2379,pd-2.tipd:2379")
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] server.notify-capacity keep default 40960
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] server.end-point-concurrency keep default 8
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] server.messages-per-tick keep default 4096
2017/03/09 13:30:54.101 tikv-server.rs:108: [INFO] flag capacity: None
2017/03/09 13:30:54.101 tikv-server.rs:97: [INFO] flag advertise-addr: Some("tikv-1.tikv:20160")
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] server.send-buffer-size keep default 131072
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] server.recv-buffer-size keep default 131072
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] raftstore.notify-capacity keep default 40960
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] raftstore.messages-per-tick keep default 4096
2017/03/09 13:30:54.101 tikv-server.rs:188: [INFO] raftstore.raft-base-tick-interval keep default 1000
2017/03/09 13:30:54.101 tikv-server.rs:176: [INFO] raftstore.raft-heartbeat-ticks keep default 2
2017/03/09 13:30:54.102 tikv-server.rs:176: [INFO] raftstore.raft-election-timeout-ticks keep default 10
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.split-region-check-tick-interval keep default 10000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.region-split-size keep default 67108864
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.region-max-size keep default 83886080
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.region-split-check-diff keep default 8388608
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.raft-log-gc-tick-interval keep default 10000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.raft-log-gc-threshold keep default 50
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.raft-log-gc-count-limit keep default 49152
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.raft-log-gc-size-limit keep default 50331648
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.region-compact-check-interval keep default 300000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.region-compact-delete-keys-count keep default 1000000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.lock-cf-compact-interval keep default 600000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.raft-entry-max-size keep default 8388608
2017/03/09 13:30:54.102 tikv-server.rs:198: [INFO] raftstore.max-peer-down-duration keep default Duration { secs: 300, nanos: 0 }
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.pd-heartbeat-tick-interval keep default 60000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.pd-store-heartbeat-tick-interval keep default 10000
2017/03/09 13:30:54.102 tikv-server.rs:188: [INFO] raftstore.consistency-check-interval keep default 0
2017/03/09 13:30:54.102 tikv-server.rs:176: [INFO] raftstore.accelerate-campaign-reserved-ticks keep default 1
2017/03/09 13:30:54.102 tikv-server.rs:176: [INFO] storage.scheduler-notify-capacity keep default 10240
2017/03/09 13:30:54.102 tikv-server.rs:176: [INFO] storage.scheduler-messages-per-tick keep default 1024
2017/03/09 13:30:54.102 tikv-server.rs:176: [INFO] storage.scheduler-concurrency keep default 102400
2017/03/09 13:30:54.102 tikv-server.rs:176: [INFO] storage.scheduler-worker-pool-size keep default 4
2017/03/09 13:30:54.102 tikv-server.rs:97: [INFO] flag labels: None
2017/03/09 13:30:54.102 tikv-server.rs:131: [INFO] server.labels use default Some("")
2017/03/09 13:30:54.102 tikv-server.rs:136: [INFO] toml value server.labels: ""
2017/03/09 13:30:54.102 tikv-server.rs:97: [INFO] flag s: Some("/data")
2017/03/09 13:30:54.103 tikv-server.rs:131: [INFO] server.backup use default Some("")
2017/03/09 13:30:54.103 tikv-server.rs:136: [INFO] toml value server.backup: ""
2017/03/09 13:30:54.103 tikv-server.rs:677: [INFO] backup path: /data/backup
2017/03/09 13:30:54.129 mod.rs:193: [INFO] starting working thread: store address resolve worker
2017/03/09 13:31:06.589 panic_hook.rs:85: [ERROR] thread 'main' panicked 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 22, message: "Invalid argument" } }' at "/buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/result.rs:868"
stack backtrace:
   0:     0x7f1f0ef917ae - backtrace::backtrace::libunwind::trace
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54
                         - backtrace::backtrace::trace<closure>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70
   1:     0x7f1f0ef91e53 - backtrace::capture::{{impl}}::new
                        at /home/jenkins/.target/release/build/backtrace-811cded9ab5b8ade/out/capture.rs:79
   2:     0x7f1f0ea3f3a7 - tikv::util::panic_hook::set_exit_hook::{{closure}}
                        at /home/jenkins/workspace/TIKV_POST_COMMIT_FLOW/go/src/github.com/pingcap/tikv/src/util/panic_hook.rs:84
   3:     0x7f1f0f0788e8 - std::panicking::rust_panic_with_hook
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:546
   4:     0x7f1f0f078764 - std::panicking::begin_panic<collections::string::String>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:507
   5:     0x7f1f0f0786d9 - std::panicking::begin_panic_fmt
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:491
   6:     0x7f1f0f078667 - std::panicking::rust_begin_panic
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:467
   7:     0x7f1f0f0a59dd - core::panicking::panic_fmt
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/panicking.rs:69
   8:     0x7f1f0e948aff - core::result::unwrap_failed<std::io::error::Error>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/macros.rs:29
   9:     0x7f1f0e9cf74c - core::result::{{impl}}::unwrap<std::fs::File,std::io::error::Error>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/result.rs:745
                         - tikv_server::run_raft_server
                        at /home/jenkins/workspace/TIKV_POST_COMMIT_FLOW/go/src/github.com/pingcap/tikv/src/bin/tikv-server.rs:719
  10:     0x7f1f0e9d7e40 - tikv_server::main
                        at /home/jenkins/workspace/TIKV_POST_COMMIT_FLOW/go/src/github.com/pingcap/tikv/src/bin/tikv-server.rs:875
  11:     0x7f1f0f07fa7a - panic_unwind::__rust_maybe_catch_panic
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libpanic_unwind/lib.rs:98
  12:     0x7f1f0f079086 - std::panicking::try<(),fn()>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panicking.rs:429
                         - std::panic::catch_unwind<fn(),()>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/panic.rs:361
                         - std::rt::lang_start
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libstd/rt.rs:57
  13:     0x7f1f0d9ba874 - __libc_start_main
  14:     0x7f1f0e8e802d - <unknown>
  15:                0x0 - <unknown>
```

--- a/src/bin/tikv-server.rs
+++ b/src/bin/tikv-server.rs
@@ -216,10 +216,11 @@ fn initial_log(matches: &Matches, config: &toml::Value) {
     let level_filter = logger::get_level_by_string(&level);
     if log_file_path.is_empty() {
         let w = StderrLogger;
-        logger::init_log(w, level_filter).unwrap();
+        logger::init_log(w, level_filter).unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     } else {
-        let w = RotatingFileLogger::new(&log_file_path).unwrap();
-        logger::init_log(w, level_filter).unwrap();
+        let w = RotatingFileLogger::new(&log_file_path)
+            .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
+        logger::init_log(w, level_filter).unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     }
 }
 
@@ -241,7 +242,7 @@ fn initial_metric(config: &toml::Value, node_id: Option<u64>) {
 
     info!("start prometheus client");
 
-    util::monitor_threads("tikv").unwrap();
+    util::monitor_threads("tikv").unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
 
     util::run_prometheus(Duration::from_millis(push_interval as u64),
                          &push_address,
@@ -280,7 +281,8 @@ fn check_advertise_address(addr: &str) {
 fn get_rocksdb_db_option(config: &toml::Value) -> RocksdbOptions {
     let mut opts = RocksdbOptions::new();
     let rmode = get_toml_int(config, "rocksdb.wal-recovery-mode", Some(2));
-    let wal_recovery_mode = util::config::parse_rocksdb_wal_recovery_mode(rmode).unwrap();
+    let wal_recovery_mode = util::config::parse_rocksdb_wal_recovery_mode(rmode)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     opts.set_wal_recovery_mode(wal_recovery_mode);
 
     let wal_dir = get_toml_string(config, "rocksdb.wal-dir", Some("".to_owned()));
@@ -395,7 +397,8 @@ fn get_rocksdb_cf_option(config: &toml::Value,
     let cpl = get_toml_string(config,
                               (prefix.clone() + "compression-per-level").as_str(),
                               Some("lz4:lz4:lz4:lz4:lz4:lz4:lz4".to_owned()));
-    let per_level_compression = util::config::parse_rocksdb_per_level_compression(&cpl).unwrap();
+    let per_level_compression = util::config::parse_rocksdb_per_level_compression(&cpl)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     opts.compression_per_level(&per_level_compression);
 
     let write_buffer_size = get_toml_int(config,
@@ -459,7 +462,7 @@ fn get_rocksdb_write_cf_option(config: &toml::Value, total_mem: u64) -> RocksdbO
     // prefix extractor(trim the timestamp at tail) for write cf.
     opt.set_prefix_extractor("FixedSuffixSliceTransform",
                               Box::new(rocksdb_util::FixedSuffixSliceTransform::new(8)))
-        .unwrap();
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     // create prefix bloom for memtable.
     opt.set_memtable_prefix_bloom_size_ratio(0.1 as f64);
     opt
@@ -477,7 +480,7 @@ fn get_rocksdb_raftlog_cf_option(config: &toml::Value, total_mem: u64) -> Rocksd
     let mut opt = get_rocksdb_cf_option(config, "raftcf", default_block_cache_size, false, false);
     opt.set_memtable_insert_hint_prefix_extractor("RaftPrefixSliceTransform",
             Box::new(rocksdb_util::FixedPrefixSliceTransform::new(region_raft_prefix_len())))
-        .unwrap();
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     opt
 }
 
@@ -495,7 +498,8 @@ fn get_rocksdb_lock_cf_option(config: &toml::Value) -> RocksdbOptions {
     opts.set_block_based_table_factory(&block_base_opts);
 
     let cpl = "no:no:no:no:no:no:no".to_owned();
-    let per_level_compression = util::config::parse_rocksdb_per_level_compression(&cpl).unwrap();
+    let per_level_compression = util::config::parse_rocksdb_per_level_compression(&cpl)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     opts.compression_per_level(&per_level_compression);
 
     let write_buffer_size = get_toml_int(config,
@@ -689,11 +693,14 @@ fn build_raftkv(config: &toml::Value,
     cfs_opts.insert(CF_RAFT, get_rocksdb_raftlog_cf_option(config, total_mem));
     let mut db_path = path.clone();
     db_path.push("db");
-    let engine =
-        Arc::new(rocksdb_util::new_engine_opt(db_path.to_str().unwrap(), db_opts, cfs_opts)
-            .unwrap());
+    let engine = Arc::new(rocksdb_util::new_engine_opt(db_path.to_str()
+                                                           .unwrap(),
+                                                       db_opts,
+                                                       cfs_opts)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err))));
 
-    let mut event_loop = store::create_event_loop(&cfg.raft_store).unwrap();
+    let mut event_loop = store::create_event_loop(&cfg.raft_store)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     let mut node = Node::new(&mut event_loop, cfg, pd_client);
 
     let mut snap_path = path.clone();
@@ -703,11 +710,13 @@ fn build_raftkv(config: &toml::Value,
                                     Some(node.get_sendch()),
                                     cfg.raft_store.use_sst_file_snapshot);
 
-    node.start(event_loop, engine.clone(), trans, snap_mgr.clone()).unwrap();
+    node.start(event_loop, engine.clone(), trans, snap_mgr.clone())
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     let router = ServerRaftStoreRouter::new(node.get_sendch(), node.id());
 
     (node,
-     create_raft_storage(router.clone(), engine.clone(), cfg).unwrap(),
+     create_raft_storage(router.clone(), engine.clone(), cfg)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err))),
      router,
      snap_mgr,
      engine)
@@ -719,9 +728,10 @@ fn canonicalize_path(path: &str) -> String {
         exit_with_err(format!("{} is not a directory!", path));
     }
     if !p.exists() {
-        fs::create_dir_all(p).unwrap();
+        fs::create_dir_all(p).unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     }
-    format!("{}", p.canonicalize().unwrap().display())
+    format!("{}",
+            p.canonicalize().unwrap_or_else(|err| exit_with_err(format!("{:?}", err))).display())
 }
 
 fn get_store_and_backup_path(matches: &Matches, config: &toml::Value) -> (String, String) {
@@ -753,7 +763,8 @@ fn get_store_and_backup_path(matches: &Matches, config: &toml::Value) -> (String
 fn get_store_labels(matches: &Matches, config: &toml::Value) -> HashMap<String, String> {
     let labels = get_flag_string(matches, "labels")
         .unwrap_or_else(|| get_toml_string(config, "server.labels", Some("".to_owned())));
-    util::config::parse_store_labels(&labels).unwrap()
+    util::config::parse_store_labels(&labels)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)))
 }
 
 fn start_server<T, S>(mut server: Server<T, S>,
@@ -767,11 +778,11 @@ fn start_server<T, S>(mut server: Server<T, S>,
     let h = thread::Builder::new()
         .name("tikv-eventloop".to_owned())
         .spawn(move || {
-            server.run(&mut el).unwrap();
+            server.run(&mut el).unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
         })
-        .unwrap();
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     signal_handler::handle_signal(ch, engine, backup_path);
-    h.join().unwrap();
+    h.join().unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
 }
 
 fn run_raft_server(pd_client: RpcClient,
@@ -779,15 +790,17 @@ fn run_raft_server(pd_client: RpcClient,
                    backup_path: &str,
                    config: &toml::Value,
                    total_mem: u64) {
-    let mut event_loop = create_event_loop(&cfg).unwrap();
+    let mut event_loop = create_event_loop(&cfg)
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     let ch = SendCh::new(event_loop.channel(), "raft-server");
     let pd_client = Arc::new(pd_client);
-    let resolver = PdStoreAddrResolver::new(pd_client.clone()).unwrap();
+    let resolver = PdStoreAddrResolver::new(pd_client.clone())
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
 
     let store_path = &cfg.storage.path;
     let mut lock_path = Path::new(store_path).to_path_buf();
     lock_path.push("LOCK");
-    let f = File::create(lock_path).unwrap();
+    let f = File::create(lock_path).unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     if f.try_lock_exclusive().is_err() {
         panic!("lock {} failed, maybe another instance is using this directory.",
                store_path);
@@ -805,7 +818,7 @@ fn run_raft_server(pd_client: RpcClient,
     }
 
     info!("Start listening on {}...", cfg.addr);
-    let listener = bind(&cfg.addr).unwrap();
+    let listener = bind(&cfg.addr).unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
 
     let server_chan = ServerChannel {
         raft_router: raft_router,
@@ -818,9 +831,9 @@ fn run_raft_server(pd_client: RpcClient,
                           server_chan,
                           resolver,
                           snap_mgr)
-        .unwrap();
+        .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     start_server(svr, event_loop, engine, backup_path);
-    node.stop().unwrap();
+    node.stop().unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
 }
 
 fn main() {
@@ -928,9 +941,9 @@ fn main() {
     let cluster_id = pd_client.get_cluster_id()
         .unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
 
-    let total_cpu_num = cpu_num().unwrap();
+    let total_cpu_num = cpu_num().unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     // return  memory in KB.
-    let mem = mem_info().unwrap();
+    let mem = mem_info().unwrap_or_else(|err| exit_with_err(format!("{:?}", err)));
     let total_mem = mem.total * KB;
     if !sanitize_memory_usage() {
         panic!("default block cache size over total memory.");

